packages/bsync/src/purchases/purchasesClient.ts

```diff
@@ -0,0 +1,269 @@
+import { Subscription } from '../proto/bsync_pb'
+import { RcSubscription, RevenueCatClient } from './revenueCatClient'
+
+type PlatformId = 'web' | 'ios' | 'android'
+
+type SubscriptionGroupId = 'core'
+
+type SubscriptionOfferingId = 'coreMonthly' | 'coreAnnual'
+
+const assertSubscriptionGroup: (
+  group: string,
+) => asserts group is SubscriptionGroupId = (group: string) => {
+  if (['core'].includes(group)) {
+    return
+  }
+  throw new Error(`invalid subscription group: '${group}'`)
+}
+
+const assertPlatform: (platform: string) => asserts platform is PlatformId = (
+  platform: string,
+) => {
+  if (['web', 'ios', 'android'].includes(platform)) {
+    return
+  }
+  throw new Error(`invalid platform: '${platform}'`)
+}
+
+export type PurchasesClientOpts = {
+  revenueCatV1ApiKey: string
+  revenueCatV1ApiUrl: string
+  revenueCatWebhookAuthorization: string
+  stripePriceIdMonthly: string
+  stripePriceIdAnnual: string
+  stripeProductIdMonthly: string
+  stripeProductIdAnnual: string
+}
+
+const mapStoreToPlatform = (store: string): PlatformId | undefined => {
+  switch (store) {
+    case 'stripe':
+      return 'web'
+    case 'app_store':
+      return 'ios'
+    case 'play_store':
+      return 'android'
+    default:
+      return undefined
+  }
+}
+
+export class PurchasesClient {
+  private revenueCatClient: RevenueCatClient
+  private STRIPE_PRODUCTS: {
+    [subscriptionOfferingId in SubscriptionOfferingId]: string
+  }
+
+  /**
+   * All of our purchaseable product IDs for all platforms.
+   */
+  private PRODUCTS: {
+    [platform in PlatformId]: {
+      [subscriptionOfferingId in SubscriptionOfferingId]: string
+    }
+  }
+
+  /**
+   * Manual groupings of products into "Subscription Groups", mimicking the way
+   * Apple does it.
+   */
+  private GROUPS: {
+    [group in SubscriptionGroupId]: {
+      [platform in PlatformId]: {
+        id: SubscriptionOfferingId
+        product: string
+      }[]
+    }
+  }
+
+  constructor(opts: PurchasesClientOpts) {
+    this.revenueCatClient = new RevenueCatClient({
+      v1ApiKey: opts.revenueCatV1ApiKey,
+      v1ApiUrl: opts.revenueCatV1ApiUrl,
+      webhookAuthorization: opts.revenueCatWebhookAuthorization,
+    })
+
+    this.STRIPE_PRODUCTS = {
+      coreMonthly: opts.stripeProductIdMonthly,
+      coreAnnual: opts.stripeProductIdAnnual,
+    }
+
+    this.PRODUCTS = {
+      web: {
+        coreMonthly: opts.stripePriceIdMonthly,
+        coreAnnual: opts.stripePriceIdAnnual,
```

Comment:
I think this is totally reasonable (see other comment about test/prod) but keep in mind: as we add more products, we'll end up with a lot of config. I'd maybe argue that it's safe to hard-code these values in here too, and add a single piece of top-level config like `isTestMode`. There may be other handling we only want in test mode too, so might be good to narrow the `opts` interface and just make this more verbose here.

Reply:
i think either is probably fine tbh. I'd be inclined to just leave it like this for now & switch out how we do it as the config gets more complex